Doc/library/sqlite3.rst

```diff
@@ -2358,6 +2330,84 @@ can be found in the `SQLite URI documentation`_.
 .. _SQLite URI documentation: https://www.sqlite.org/uri.html
 
 
+.. _sqlite3-row-factory-how-to:
+
+How to work with row factories
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` represent fetched rows as :class:`tuples <tuple>`.
+If a :class:`!tuple` does not suit your needs,
+use the built-in :class:`Row` type or a custom :attr:`~Connection.row_factory`.
+The former provides both indexed and case-insensitive named access to columns,
+with low memory overhead and minimal performance impact.
+Example use:
+
+.. doctest::
+
+   >>> con = sqlite3.connect(":memory:")
+   >>> con.row_factory = sqlite3.Row
+   >>> res = con.execute("SELECT 'Earth' AS name, 6378 AS radius")
+   >>> row = res.fetchone()
+   >>> row.keys()
+   ['name', 'radius']
+   >>> row[0], row["name"]  # Access by index and name.
+   ('Earth', 'Earth')
+   >>> row["RADIUS"]  # Column names are case-insensitive.
+   6378
+
+If you need more flexibility, you can design your own row factory.
+Here's an example of a :class:`dict` row factory:
+
+.. doctest::
+
+   >>> def dict_factory(cursor, row):
+   ...     col_names = [col[0] for col in cursor.description]
+   ...     return {key: value for key, value in zip(col_names, row)}
+
+   >>> con = sqlite3.connect(":memory:")
+   >>> con.row_factory = dict_factory
+   >>> for row in con.execute("SELECT 1 AS a, 2 AS b"):
+   ...     print(row)
+   {'a': 1, 'b': 2}
+
+Here's an example of a optimised :class:`~collections.namedtuple` factory:
+
+.. testcode::
+
+   from collections import namedtuple
+   from functools import lru_cache
+
+   def _fields(cursor):
+        return (col[0] for col in cursor.description)
+
+   @lru_cache
+   def _make_cls(fields):
+       return namedtuple("Row", fields)
+
+   def NamedTupleFactory(cursor, row):
+       cls = _make_cls(_fields(cursor))
+       return cls._make(row)
```

Comment:
Is it not possible to avoid using a "private" API of `namedtuple` here?

Reply:
AFAICS, that's the recommended API: https://docs.python.org/3/library/collections.html#collections.somenamedtuple._make

Reply:
Oh, I see—it was named that way to avoid clashing with field names, I presume, though modern convention for this would be using an underscore at the end (`.make_`) rather than the beginning, to avoid conflating it with the convention for "private".